openassessment/workflow_batch_update_api.py

```diff
@@ -0,0 +1,595 @@
+
+        Sample structure (assessment_requirements and course_settings content omitted):
+        ```
+        {
+          "courses": [
+            {
+              "course_id": "value",
+              "course_settings": {},
+              "assessments": [
+                {
+                  "item_id": "value",
+                  "assessment_requirements": {},
+                  "submissions": [
+                    {
+                      "submission_uuid": "value"
+                    }
+                  ]
+                }
+              ]
+            }
+          ]
+        }
+    ```
+
+    Raises:
+        OraWorkflowBatchUpdateException: If batch process fails and cannot continue, e.g. number of errors
+                                        threshold was exceeded, etc.
+    """
+    try:
+        start = time.time()
+
+        if workflow_update_parameters is None:
+            peer_workflows = get_blocked_peer_workflows(item_id=item_id)
+            workflow_update_parameters = get_workflow_update_parameters(peer_workflows)
+
+        course_settings = _get_course_settings_for_ora(workflow_update_parameters, item_id)
+        workflow_update_parameters_for_ora = _get_ora_object(workflow_update_parameters, item_id)
+
+        if workflow_update_parameters_for_ora is not None and workflow_update_parameters_for_ora.get(
+                'assessment_requirements') is not None:
+            assessment_requirements = workflow_update_parameters_for_ora['assessment_requirements']
+
+            for item in workflow_update_parameters_for_ora["submissions"]:
+                # execute asynchronously (submit Celery task)
+                update_workflow_for_submission_task.apply_async(
+                    [item["submission_uuid"], assessment_requirements, course_settings])
+
+            end = time.time()
+            logger.info(
+                "Batch workflow update for blocked ORA submissions completed successfully. "
+                "item_id=%s assessment_requirements=%s course_settings=%s submissions=%s processing_time=%s",
+                item_id,
+                assessment_requirements,
+                course_settings,
+                workflow_update_parameters_for_ora["submissions"],
+                str(end - start))
+        else:
+            logger.info("No blocked ORA submissions found for the ORA item_id=%s . "
+                        "Batch workflow update completed without submitting any tasks.",
+                        item_id)
+
+    except (UpdateWorkflowForSubmissionException, Exception) as e:  # pylint: disable=broad-except
+        logger.error(
+            "Batch workflow update for ORA block failed. "
+            "item_id=%s error_message:%s",
+            item_id,
+            str(e))
+        raise UpdateWorkflowsForOraBlockException(str(e)) from e
+
+
+def update_workflow_for_submission(submission_uuid, assessment_requirements, course_settings):
```

Comment:
Again, I think this could be helpful to have a version where `assessment_requirements` and `course_settings` are `kwargs`. That way, you could call this flexibly, without having to do those lookups in advance and having this function work better for that scope, but omitting those searches when the data is provided.
```suggestion
def update_workflow_for_submission(submission_uuid, assessment_requirements=None, course_settings=None):
```

Reply:
Do we ever envision running asynchronous job for as single submission?

Reply:
This is not the asynchronous job, this is an API util, no? There's a separate wrapper for the async job.

The challenge of keeping it like this (and including things like timing scripts, etc. in this scope) is you are limiting yourself to one way of using this API utility. If you ever **do** want to do something similar, but with slight variation, you're going to have to make new versions of this function.

In general, the tasks can be pretty rigid but it's useful to have the underlying APIs be pretty general-purpose.

Reply:
To end the thread, I have made updates for this, all update_* signatures are consolidated for similar behavior.